### PR TITLE
244 3.27 buttons according to event states

### PR DESCRIPTION
**Pull request by ekrebs5**
This PR is part of #244 
It only focuses on the visibility of the buttons mentioned in the user story (Task: "buttons for print all, accept all, sending emails and printing badges and viewing the participant list are viewable dependent on state"). All the buttons are on the page of an event.

| button \ phase         | draft   | application | selection | execution |
|----------------------|----|-------------|-----------|-----------|
| print all                    |  no     | no          | yes       | no        |
| accept all               | no  | no          | yes       | no        |
| send accept/reject email | no   | no         | yes       | no        |
| print badges             | no     | no       | no        | yes       |
| participant list         | no     | no       | no        | yes       |

EDIT:
print badges is on the participant list site and its visibility is thus dependent on the participant list button

### CODE REVIEW

File: spec/views/events/show.html.erb_spec.rb [hunk 1]
```diff
@@ -54,17 +54,8 @@
...
-
-  it "displays print badges button" do
-    render
-    expect(rendered).to have_link(t(:print_button_label, scope: 'events.badges'))
-  end
-
   it "should not display accept-all-button for non-organizers" do
+    @event = assign(:event, FactoryGirl.create(:event, :in_selection_phase))
     @event.max_participants = Float::INFINITY
     [:coach, :student].each do | each |
```
Comment (aloeser) [inline spec/views/events/show.html.erb_spec.rb]: not your change... but since when do we have :student as a role?

File: spec/views/events/show.html.erb_spec.rb [hunk 2]
```diff
@@ -54,17 +54,8 @@
     expect(rendered).to have_link(t(:details, scope: 'events.applicants_overview'))
   end
 
-  it "displays print applications button" do
```
Comment (aloeser) [inline spec/views/events/show.html.erb_spec.rb:57]: why did you delete this test?

File: spec/views/events/show.html.erb_spec.rb [hunk 3]
```diff
@@ -54,17 +54,8 @@
...
-    render
-    expect(rendered).to have_link(t(:print_all, scope: 'events.applicants_overview'))
-  end
-
-  it "displays print badges button" do
```
Comment (aloeser) [inline spec/views/events/show.html.erb_spec.rb]: why did you delete this test?

**Review by aloeser — COMMENTED** [review]
It seems like feature tests for print badges and print application buttons are missing?
Also it might be a good idea to add a test, that when in application phase, even with pending applications or too many accepted applications the buttons are not visible. This seemed buggy for me, and it would be nice to have a test for it

**Comment by aloeser** [comment]
@tobiduer should it be possible to see the event details page for an event in draft phase?
In that case, we would need some more tests

File: spec/views/events/show.html.erb_spec.rb (hunk 2 above)
Comment (ekrebs5) [reply spec/views/events/show.html.erb_spec.rb:57]: this is now event phase dependent and covered in a later test
(as in this test would be duplicate with the exact lines)

File: spec/views/events/show.html.erb_spec.rb (hunk 3 above)
Comment (ekrebs5) [reply spec/views/events/show.html.erb_spec.rb]: this is now event phase dependent and covered in a later test
(as in this test would be duplicate with the exact lines)

**Comment by ekrebs5** [comment]
@aloeser 
"It seems like feature tests for print badges and print application buttons are missing?"
Could be, not our issue, so you should make a technical debt issue

"Also it might be a good idea to add a test, that when in application phase, even with pending applications or too many accepted applications the buttons are not visible. This seemed buggy for me, and it would be nice to have a test for it"
They are added now

File: spec/views/events/show.html.erb_spec.rb (hunk 1 above)
Comment (ekrebs5) [reply spec/views/events/show.html.erb_spec.rb]: I don't know why that is there, should I change it to pupil? (looks like this was somehow not renamed, I think in the very beginning there was a student)

File: app/views/events/_applicants_overview.html.erb [hunk 4]
```diff
@@ -57,36 +57,27 @@
...
         <%= link_to t('.accept_all'), event_accept_all_applicants_path(@event.to_param), :class => 'btn btn-default' %>
       <% end %>
-      <%= link_to t('.print_all'), print_applications_event_path(@event.to_param), :target => "_blank", :class => 'btn btn-default' %>
+      <% if @event.phase == :selection %>
+        <%= link_to t('.print_all'), print_applications_event_path(@event.to_param), :target => "_blank", :class => 'btn btn-default' %>
+      <% end %>
     </div>
 
-    <% if not @event.applications_classified? %>
-      <div class="btn-group pull-right tooltip-wrapper has-tooltip" role="group" data-toggle="tooltip" title="<%= t '.unclassified_applications_left'%>">
```
Comment (aloeser) [inline app/views/events/_applicants_overview.html.erb:67]: you loose this line

File: app/views/events/_applicants_overview.html.erb [hunk 5]
```diff
@@ -57,36 +57,27 @@
...
-        <%= link_to t('events.badges.print_button_label'), event_badges_path(@event), :class => 'btn btn-default'%>
-        <% if can? :send_email, Email %>
-          <%= button_tag t('.sending_acceptances'), type: 'button', class: 'send-emails-button btn btn-default', disabled: true %>
-          <%= button_tag t('.sending_rejections'), type: 'button', class: 'send-emails-button btn btn-default', disabled: true %>
-        <% end %>
-      </div>
-    <% elsif (@event.compute_free_places < 0) %>
-      <div class="btn-group pull-right tooltip-wrapper has-tooltip" role="group" data-toggle="tooltip" title="<%= t '.maximum_number_of_participants_exeeded'%>">
```
Comment (aloeser) [inline app/views/events/_applicants_overview.html.erb:74]: you loose this line

File: spec/views/events/show.html.erb_spec.rb [hunk 6]
```diff
@@ -101,4 +89,64 @@
...
+    render
+    expect(rendered).to_not have_link(t(:print_all, scope: 'events.applicants_overview'))
+    expect(rendered).to_not have_link(t(:accept_all, scope: 'events.applicants_overview'))
+    expect(rendered).to_not have_link(t(:sending_acceptances, scope: 'events.applicants_overview'))
+    expect(rendered).to_not have_link(t(:sending_rejections, scope: 'events.applicants_overview'))
+    expect(rendered).to_not have_link(t(:print_button_label, scope: 'events.badges'))
+    expect(rendered).to_not have_link(t(:show_participants, scope: 'events.participants'))
+  end
+
+  it "does not display the disabled send email buttons in application phase" do
```
Comment (aloeser) [inline spec/views/events/show.html.erb_spec.rb]: I suggest to append something like "despite unclassfied applications", to make clear what edge case you are testing.
Test itself looks good

**Review by aloeser — CHANGES_REQUESTED** [review]
You need to give  app/views/events/_applicants_overview.html.erb another glance.
I think the two lines you deleted killed the tooltip on the send accept/reject mail buttons (which tell you that you cant send mails now, since too many participants/unclassified remain)

**Comment by tobiasduerschmid** [comment]
@aloeser: Yes, but only for organizers

**Review by aloeser — APPROVED** [review]

File: spec/views/events/participants.html.erb_spec.rb [hunk 7]
```diff
@@ -49,17 +49,42 @@
...
 
   it "disables the print badges button when there are no participants" do
+    @event = assign(:event, FactoryGirl.create(:event, :in_execution_phase))
     @participants = []
     render
     expect(rendered).to have_link(t(:print_button_label, scope: 'events.badges', disabled: true))
   end
 
+  it "does not show the print badges button when the event is in draft phase" do
+    @event = assign(:event, FactoryGirl.create(:event_with_accepted_applications, :in_draft_phase))
```
Comment (querenker) [inline spec/views/events/participants.html.erb_spec.rb:68]: Is it even possible to have accepted application in the draft phase?

File: spec/views/events/show.html.erb_spec.rb [hunk 8]
```diff
@@ -96,4 +89,78 @@
...
+    sign_in(FactoryGirl.create(:user, role: :organizer))
+    render
+    expect(rendered).to have_link(t(:print_all, scope: 'events.applicants_overview'))
+    expect(rendered).to have_link(t(:accept_all, scope: 'events.applicants_overview'))
+    expect(rendered).to have_link(t(:sending_acceptances, scope: 'events.applicants_overview'))
+    expect(rendered).to have_link(t(:sending_rejections, scope: 'events.applicants_overview'))
+    expect(rendered).to_not have_link(t(:show_participants, scope: 'events.participants'))
+  end
+
+  it "does display the disabled send email buttons in selection phase (when there are unclassified applications)" do
```
Comment (querenker) [inline spec/views/events/show.html.erb_spec.rb]: Just for consistency. In the other cases it is "it displays..." instead of "it does display...."

File: spec/views/events/participants.html.erb_spec.rb (hunk 7 above)
Comment (ekrebs5) [reply spec/views/events/participants.html.erb_spec.rb:68]: no, it isn't
I added new tests just now because the button was moved
is the participant page only visible when there are accepted participants?

File: spec/views/events/show.html.erb_spec.rb [hunk 9]
```diff
@@ -96,4 +89,78 @@
...
+    @event = assign(:event, FactoryGirl.create(:event, :with_diverse_open_applications, :in_selection_phase))
+    sign_in(FactoryGirl.create(:user, role: :organizer))
+    render
+    expect(rendered).to have_button(t(:sending_acceptances, scope: 'events.applicants_overview'), disabled: true)
+    expect(rendered).to have_button(t(:sending_rejections, scope: 'events.applicants_overview'), disabled: true)
+  end
+
+  it "does display the disabled send email buttons in selection phase (when there are too many accepted applications)" do
```
Comment (querenker) [inline spec/views/events/show.html.erb_spec.rb]: same as above

**Review by querenker — APPROVED** [review]
Just some minor comments

File: spec/views/events/participants.html.erb_spec.rb (hunk 7 above)
Comment (ekrebs5) [reply spec/views/events/participants.html.erb_spec.rb:68]: wait, you mean like draft and selection are a subset of when they are no participants?

I think these tests should still be there, because in these cases there should be no button at all, while in the "normal" case with no participants it's disabled
